### ISSUE TITLE
[RFC] TUI interface for summoner

### ISSUE THREAD

**Issue by vrom911** [kowainik/summoner#187]
# Proposal
Looking at the amount of the questions in the interactive mode of `summoner` I feel a bit unsatisfied. To get rid of that I propose to make `summoner` a terminal app with TUI. So you will see all settings on one screen where you'll be able to select the checkboxes  or to edit names/other text options like in CLI that you'd like to see in the generated project.

Note, that it should work absolutely identical to what we have now with the only difference – `summoner` appearance (something like this very rough sketch):

```
┌─────────────  Summoner  ───────────────┐
│                                        │
│ Project name: [                ]       │
│                                        │
│                    ...                 │
│                                        │
│┌───────────  Build Tool  ─────────────┐│
││[x] Cabal                             ││
││[x] Stack                             ││
││[ ] Nix                               ││
│└──────────────────────────────────────┘│
│                ...                     │
└────────────────────────────────────────┘
```

We can use [`brick`]() librarary for this.

## Pros

* More flexible interface
* Ability to see all settings you can change immediatly
* On-fly check of invalid fields
* Easy edit of last steps (which is not possible in CLI, see #177)
* Possibility to easy add features (just use your imagination)

## Cons

Unfortunately, this feature has some drawbacks. 

* __Important:__ Windows support
  _Solution:_ It's not good to just stop supporting this OS, so the solution is to have `summoner-cli` and `summoner-tui` as internal libraries, so anyone can at least install the tool. Also, there will be two executables: `summon` and `summon-tui` (or any other names), the first one will depend on the main `summoner-cli` library (which is the one we have now) and the second one will run TUI interface for summoner. 
* Few internal libraries to maintain
* Complicated installation process (Use some flags for installing the certain library (CLI or TUI))
* `brick` is quite heavy dependency

---

Basically, we are not going to give up everything we have now, but will add feature with more robust interface for users who is comfortable with it.

Any comments, thought, other drawbacks are appreciated.

**Comment by ChShersh** [comment]
I really like the idea :heart_eyes: I see that there're a lot of crazy possibilities that we can do with TUI:

1. Display project directory content on-the-fly as you change settings.
2. Add possibility to view or even change content of the generated files (for example, if you want to edit `.cabal` file a little bit or `.travis.yml`).

And much more. I expect that it will require a lot of time to implement and support all features, but I really like the idea :100: :top:

**Comment by ChShersh** [comment]
@willbasky It's possible but it's much more difficult. Requires to dance with `-XCPP` a lot so my guess is that we're not going to follow this way. Let's see how people will like the idea of TUI for `summoner` and maybe in some future we will release `summoner-2.0` where TUI is the default behavior. 

Also, when talking about single executable, it's better to have `--tui` and `--cli` as flags instead of commands (keeping one of them as default choice) because it's more backwards compatible.

**Comment by qoelet** [comment]
I like the idea, especially with the on-the-fly checking. @willbasky @ChShersh Having CLI flags to toggle between the two modes sounds perfect. Personally, I would vote for TUI as the default one, with `--cli` or `--plain` to force plain ol' CLI.
